### DISCUSSION TITLE
is there a change log for the api?  Trying to find when /observations/current got depricated

### DISCUSSION THREAD

**Discussion by edclow** [weather-gov/api#482] [Q&A]
Seems like observations/current has been either deprecated and/or replaced with observations/latest

If this accurate?

**Comment by akirmse** [comment]
Seconding this request, as well as a place to go for announcements a few months before breaking changes are made.  I just found out my 50k mobile users are broken because the API changed recently.

**Comment by scadergit** [comment]
This endpoint was deprecated in the API and noted in the specification over a year ago, but then the deploy schedule went different than expected and the notice of removal was lost in the shuffle. We recognize the importance of a service to notify customers of breaking changes.

The official notice would have been an [NWS Service Change Notice (SCN)](https://www.weather.gov/notification/), but many developers don't know those exist let alone monitor them for changes to the API. We have debated other solutions to compliment SCN's, but we haven't formally selected one. Suggestions are welcome.

**Comment by akirmse** [comment]
Thanks for the quick response.  Agreed, the SCN mechanism looks... unwieldy.  What I've seen other API owners do is send out announcements on a low-volume, opt-in mailing list.  Another option, not quite as good IMO, is a blog that could be monitored by RSS or similar.

(BTW, the API I was using was https://api.weather.gov/points/XXX,YYY/forecast.  It looks like I now need to do two round-trips to get the same result.  A really good API design would continue to do this work once, internally, rather than having thousands of identical implementations in clients.  Unless there's a direct replacement that I didn't see.)

**Reply by scadergit** [reply]
It's been a long-time goal to implement API keys, which would inherently make mailing lists more feasible.

**Reply by StephenClouse** [reply]
> BTW, the API I was using was https://api.weather.gov/points/XXX,YYY/forecast. It looks like I now need to do two round-trips to get the same result. A really good API design would continue to do this work once, internally, rather than having thousands of identical implementations in clients.

This is how we implemented it originally, as evidenced by the fact that the old endpoint existed. But our design concerns tend towards cacheability and edge offload (which in turn have a dramatic effect on performance). The thing with gridpoints is that the NWS operates on a 2.5-km grid, so every lat/lon point in that square would be returning the exact same data. The result of that was terabytes of duplicate cached data that was not offloading (because a tiny lat/lon change would become a cache miss) when it easily could. This was not easily solvable as long as lat/lon was the authoritative mapping. This is why we decided to push the point-to-gridpoint mapping lookup onto the consuming client.

**Comment by StephenClouse** [comment]
Now that our release schedule is un-stuck, we should be able to be more proactive in announcing these changes. This will certainly be one of the forums where we post them.

**Comment by akirmse** [comment]
Good news on announcing changes in the future.

On this particular API change, I can appreciate the need for caching for reasonable performance.  But an extra network round trip for all clients is such an impact to the user experience that I think it's worth spending considerable effort to avoid it.  On mobile clients in particular, the extra round trip is always visible to the end user.

I thought of two ways you could avoid the extra round trip and still preserve the cache hit rate:

- Convert the lat/lng to a grid ID inside an off-the-shelf caching proxy that you run, and use the grid ID as the cache key.  This might be hard to install at edge caches however.

- Ask clients to do the lat/lng to grid ID conversion internally without a network request, and publish the conversion algorithm.  You could add a new API that accepts grid IDs so that you can preserve the existing API for those who don't want to do the conversion and are willing to accept the latency of the additional round trip.

It would be best if the grid IDs used for such an API were stable and global, rather than depending on NWS office IDs, which I presume are subject to change, and are of little interest to most clients.  For example, a 2.5 km grid overlaid over your whole service area, where the grid IDs are the coordinates of the NW corner.  But even if the grid had to be more complex than this (Maybe different grids for Hawaii and Puerto Rico?  Just guessing at possible wrinkles), I think many clients would still be willing to implement it to save the round trip.  I imagine that users could provide reference implementations for most programming languages that you could host on your github site, as long as the conversion function is stable over time.

**Reply by scadergit** [reply]
There are plenty of solutions to provide that service, but at the time NCO has not allocated the compute for it. Perhaps @jah1007 can provide the official way to request new features to the API.

**Reply by StephenClouse** [reply]
> On this particular API change, I can appreciate the need for caching for reasonable performance. But an extra network round trip for all clients is such an impact to the user experience that I think it's worth spending considerable effort to avoid it. On mobile clients in particular, the extra round trip is always visible to the end user.

This is why we declare the /points endpoint as heavily cacheable. We don't think it's onerous to ask clients to utilize facilities already present in HTTP to solve this exact problem.

> It would be best if the grid IDs used for such an API were stable and global, rather than depending on NWS office IDs, which I presume are subject to change, and are of little interest to most clients. For example, a 2.5 km grid overlaid over your whole service area, where the grid IDs are the coordinates of the NW corner.

But the grids are only defined at the NWS office level. There is no global grid that is used for all offices. You make a lot of assumptions about the data that simply aren't valid.

There is no simple "algorithm" for determining these. Take, for example, this intersection of four grids (EAX, TOP, ICT, SGF):

![image](https://user-images.githubusercontent.com/468025/138612366-01d91432-5431-40a9-bb5a-d62a5084c0a9.png)

There is a certain amount of overlap between them and it is not necessarily a uniform straight line.

Additionally, office grids are rarely so well aligned as shown above. See here (MQT, APX):

![image](https://user-images.githubusercontent.com/468025/138612584-7deef3c3-a9da-43ac-883b-8a35596e2b74.png)

One needs, at minimum, the AWIPS shapefiles to properly determine the correct grid to use. This is likely beyond what you would want to pull into a mobile application, plus these change regularly and would need to be updated, plus there are still other considerations (the mere existence of a gridpoint does not signal that it is actually valid for that particular office). Resolving all of this is the service we already provide via the /points endpoint.

Changing how this functions is simply not something we are going to consider unless something significant changes with how the NWS organization overall structures its geodata.

**Reply by akirmse** [reply]
Thanks for all the detail. That really helps to explain why it currently works this way.

In an ideal API, clients wouldn't need to know or care about the distribution of NWS offices.  They would simply ask for the forecast at a point and get it in a single request.  /points might be cacheable in some applications, but in others (where users click on a map or requests are highly distributed for example) it's not. 

"Resampling" forecasts into a regular grid sounds like an interesting research project.  I'm sure a more uniform geographical representation would be very useful inside the NWS, too.

**Reply by scadergit** [reply]
In that scenario, what would you consider a low enough latency that would exonerate having to make two requests?

**Reply by akirmse** [reply]
My app is used by hikers, usually out on 3G/4G connections, and even with zero server latency, the extra round trip is noticeable.

However, I think we're quite far from that scenario today.  I have the timeout set to 20 seconds per request and regularly see periods where the request times out on wifi.  If you're able to get the tail latency down to, say, 500ms@95% I would be very happy.  But if a request times out 30% of the time (guessing---it comes and goes, and failures seem to be clustered over certain hours or days), then an extra round trip is hard to swallow.

**Reply by scadergit** [reply]
500ms was always a goal, right now we average 2-3s (outside of the temporary issues noted). We're a very small team and we've optimized the API for the current performance requirements. There are other customers (like ships at sea) that need improved latency as well. It is something we'll continue to work on, but the list of priorities continually evolves.

As an avid hiker, I expect to have no signal and look ahead at the weather, but as more towers extend the range farther out, it would be nice to have updated weather information.

**Reply by jerryyhom** [reply]
@StephenClouse, thanks for your nice explanations.

Since grids overlap, as you showed above, then when using the `points` endpoint, which gridpoint is returned?

From the documentation, precisely one gridpoint is returned.  I am guessing that overlapping gridpoints from different NWS offices might have different computed data if the sources are different.  On the other hand, the source data may be similar enough that the computed data is close enough, so overlapping gridpoints are roughly equivalent.  Even if I had a choice of gridpoints, what selection criteria would I use?  Does it matter?

**Reply by StephenClouse** [reply]
> Since grids overlap, as you showed above, then when using the `points` endpoint, which gridpoint is returned?

Well, the correct one :)

If there is more than one gridpoint within range, then we lookup the WFO actually responsible for that point via the [AWIPS basemaps](https://www.weather.gov/gis/AWIPSShapefiles).

Despite the grids having overlap, the data should not necessarily be trusted for areas that are not actually within a WFO's area of responsibility. This is why we strictly map lat/lon to gridpoint one-to-one.

**Reply by DemiMarie** [reply]
@akirmse: Would adding a bounding box to the `points` response be helpful?  @scadergit: Is this something you could provide?

A human user clicking on a map or moving around in a forest is rarely going to request the same location twice, so the ability to cache location ⇒ gridpoint mappings is not very useful.  However, being able to cache grid **zones** could be useful.  The returned bounding box does not need to be exact, so long as it is a subset of the actual bounding box.  If the new location is in the bounding box of a previously fetched gridpoint, the client can reuse the cached version.

**Reply by scadergit** [reply]
The current caching solution is to manage high-traffic requests--locations within populated areas that hit the same gridpoints. Wilderness lookups will unlikely take advantage of cache, but that is okay on our end. I'm unsure how you define a zone--are you saying you want to cache the path a hike would take? There's nothing preventing an application from capturing all the grids within a app-specific defined area to store for future use.

**Comment by akirmse** [comment]
A bounding box would be better than nothing, as I could cache it for a time, though if this assignment is known in advance, I'd much rather fetch the mapping for the whole country periodically and avoid the /points call entirely.

BTW in my app almost all weather requests are for a fixed set of mountaintop locations, so caching locations (say, on small S2 cells) would be of some use.

For what it's worth, I am still seeing timeout failures on a regular basis.  I've started using a commercial API outside the US, and it always returns in under 1 second, using a single HTTPS call.

**Reply by scadergit** [reply]
I'm unsure if you're asking us to cache grids at remote locations to reduce latency, or if you're wanting a singular forecast for an area larger than a grid?

**Reply by akirmse** [reply]
What I'd like is getting a forecast with a single request on an amortized basis.  Making two requests is a performance killer when the server is as slow as it is currently.

I think DemiMarie's suggestion was to return a polygon (or inscribed bounding box if that is impractical) with the /points response, and a way of converting lat/lng to grid coordinates inside the polygon, in the hopes that future requests from the same client are likely to lie inside this polygon and could then skip subsequent /points requests.

**Reply by DemiMarie** [reply]
Indeed it was @akirmse.

**Reply by scadergit** [reply]
Returning the forecast per point would cache less and be slower (assuming it is a service response time issue and not latency). This doesn't imply that is our preference, just the state of the infrastructure. We have a few plans to increase the response time, just need resources to implement them. The current focus is replacing the backend service that often throws errors on any given forecast request and feeds the API bad data.

We're on the same page with @DemiMarie, just today that is on your side to calculate and implement, which admitted would not be easy. If implemented on the API, it would likely be a separate endpoint that converts shapes (vs points) to a gridpoint collection, or maybe replace /points with /gridpoint/lookup that accepts multiple geometry types.

**Reply by JJKraw** [reply]
> I think DemiMarie's suggestion was to return a polygon (or inscribed bounding box if that is impractical) with the /points response, and a way of converting lat/lng to grid coordinates inside the polygon, in the hopes that future requests from the same client are likely to lie inside this polygon and could then skip subsequent /points requests.

If I'm understanding the ask correctly, if you subsequently fetch the forecast (or hourly or raw data) using the link(s) returned from the /points call, those endpoints do return a polygon within the geometry field. So if you were to cache that polygon (associating it with OfficeCode/x/y), you'd have the info you need to skip the /points call for any other location covered by that grid.

If I'm not understanding correctly, just ignore me. :)

**Reply by DemiMarie** [reply]
> The current focus is replacing the backend service that often throws errors on any given forecast request and feeds the API bad data.

Is that the source of the 500 errors I have been getting?

**Reply by scadergit** [reply]
@JJKraw yes, but it's just the geom for that one grid. I believe @DemiMarie wants to cache multiple grids to avoid future /point requests in a defined area.

**Reply by scadergit** [reply]
@DemiMarie possibly, would need the correlation id to investigate, as there are a few bugs we are currently squashing.

**Reply by DemiMarie** [reply]
Correlation ID is 5346067f

**Reply by scadergit** [reply]
@DemiMarie I searched our logging system and can't find that id. If you keep getting them, please send an email to the helpdesk.

**Reply by DemiMarie** [reply]
Thanks!
